<a href="https://www.kaggle.com/code/kennethnjuguna/ibm-dv0101en-lab?scriptVersionId=137956234" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/immigration-to-canada/canadian_immegration_data.csv


The Dataset: Immigration to canada from 1980-2013

# <h3>Import the required libraries to be used in the Notebook.</h3>

In [2]:
!pip install --upgrade numpy==1.23.0

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.0/17.0 MB 42.8 MB/s eta 0:00:0000:0100:01
  Attempting uninstall: numpy
    Found existing installation: numpy 1.23.5
    Uninstalling numpy-1.23.5:
      Successfully uninstalled numpy-1.23.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
apache-beam 2.46.0 requires dill<0.3.2,>=0.3.1.1, but you have dill 0.3.6 which is incompatible.
chex 0.1.81 requires numpy>=1.25.0, but you have numpy 1.23.0 which is incompatible.
momepy 0.6.0 requires shapely>=2, but you have shapely 1.8.5.post1 which is incompatible.
pymc3 3.11.5 requires numpy<1.22.2,>=1.15.0, but you have numpy 1.23.0 which is incompatible.
pymc3 3.11.5 requires scipy<1.8.0,>=1.7.3, but you have scipy 1.11.1 which is incompatible.
ydata-profiling 4.3.1 requires scipy<1.11,>=1.4.1, but you have scipy 1.11.1 which is incompatible.


In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import scipy as sp

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# <h3>import the required dataset canadian immigration.</h3>

In [4]:
canadaimmigration=pd.read_csv('/kaggle/input/immigration-to-canada/canadian_immegration_data.csv')

Let's view the top 5 rows of the dataset using the head() function.

In [5]:
canadaimmigration.head()

Country Continent           Region             DevName  1980  1981  \
0     Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1         Albania    Europe  Southern Europe   Developed regions     1     0   
2         Algeria    Africa  Northern Africa  Developing regions    80    67   
3  American Samoa   Oceania        Polynesia  Developing regions     0     1   
4         Andorra    Europe  Southern Europe   Developed regions     0     0   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   
3     0     0     0     0  ...     0     1     0     0     0     0     0   
4     0     0     0     0  ...     0     1     1     0     0     0     0   

   2012  2013  Total  
0  2635  2004  58639  
1   620   603  15699  
2  3774  4331  69439  
3     0     0      6  
4     1     1     15  

[5 rows x 39 columns]

We can also view the bottom 5 rows of the dataset using the tail() function.

In [6]:
canadaimmigration.tail()

Country Continent              Region             DevName  1980  \
190        Viet Nam      Asia  South-Eastern Asia  Developing regions  1191   
191  Western Sahara    Africa     Northern Africa  Developing regions     0   
192           Yemen      Asia        Western Asia  Developing regions     1   
193          Zambia    Africa      Eastern Africa  Developing regions    11   
194        Zimbabwe    Africa      Eastern Africa  Developing regions    72   

     1981  1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  \
190  1829  2162  3404  7583  5907  ...  1852  3153  2574  1784  2171  1942   
191     0     0     0     0     0  ...     0     1     0     0     0     0   
192     2     1     6     0    18  ...   161   140   122   133   128   211   
193    17    11     7    16     9  ...    91    77    71    64    60   102   
194   114   102    44    32    29  ...   615   454   663   611   508   494   

     2011  2012  2013  Total  
190  1723  1731  2112  97146  
191     0     0     0      2  
192   160   174   217   2985  
193    69    46    59   1677  
194   434   437   407   8598  

[5 rows x 39 columns]

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the info() method.

In [7]:
canadaimmigration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 39 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Country    195 non-null    object
 1   Continent  195 non-null    object
 2   Region     195 non-null    object
 3   DevName    195 non-null    object
 4   1980       195 non-null    int64 
 5   1981       195 non-null    int64 
 6   1982       195 non-null    int64 
 7   1983       195 non-null    int64 
 8   1984       195 non-null    int64 
 9   1985       195 non-null    int64 
 10  1986       195 non-null    int64 
 11  1987       195 non-null    int64 
 12  1988       195 non-null    int64 
 13  1989       195 non-null    int64 
 14  1990       195 non-null    int64 
 15  1991       195 non-null    int64 
 16  1992       195 non-null    int64 
 17  1993       195 non-null    int64 
 18  1994       195 non-null    int64 
 19  1995       195 non-null    int64 
 20  1996       195 non-null    int64

This method can be used to get a short summary of the dataframe.

In [8]:
canadaimmigration.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 39 entries, Country to Total
dtypes: int64(35), object(4)
memory usage: 59.5+ KB


Get the list of column headers we can call upon the data frame's columns instance variable.

In [9]:
canadaimmigration.columns

Index(['Country', 'Continent', 'Region', 'DevName', '1980', '1981', '1982',
       '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990', '1991',
       '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000',
       '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009',
       '2010', '2011', '2012', '2013', 'Total'],
      dtype='object')

Get the list of indices we use the .index instance variables.

In [10]:
canadaimmigration.index

RangeIndex(start=0, stop=195, step=1)

To get the index and columns as lists, we can use the tolist() method.

In [11]:
canadaimmigration.columns.tolist()

['Country',
 'Continent',
 'Region',
 'DevName',
 '1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013',
 'Total']

In [12]:
canadaimmigration.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [13]:
print(type(canadaimmigration.columns.tolist()))
print(type(canadaimmigration.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the shape instance variable of it.

In [14]:
# size of dataframe (rows, columns)
canadaimmigration.shape 

(195, 39)

# Cleaning the data

Let's clean the data set to remove a few unnecessary columns. We can use pandas drop() method as follows:

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:

In [15]:
canadaimmigration['Total'] = canadaimmigration.sum(axis=1)

/tmp/ipykernel_32/4147747809.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  canadaimmigration['Total'] = canadaimmigration.sum(axis=1)


We can check to see how many null objects we have in the dataset as follows:

In [16]:
canadaimmigration.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

# pandas Intermediate: Indexing and Selection (slicing)

In [17]:
canadaimmigration.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.

In [18]:
canadaimmigration[['Country', '1980','1981','1982','1983','1984','1985']]

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

Select Row
There are main 2 ways to select rows:

    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using set_index() method.

In [24]:
canadaimmigration.set_index('Country', inplace=True)

In [25]:
canadaimmigration.head(3)

Continent           Region             DevName  1980  1981  1982  \
Country                                                                        
Afghanistan      Asia    Southern Asia  Developing regions    16    39    39   
Albania        Europe  Southern Europe   Developed regions     1     0     0   
Algeria        Africa  Northern Africa  Developing regions    80    67    71   

             1983  1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  \
Country                              ...                                       
Afghanistan    47    71   340   496  ...  3436  3009  2652  2111  1746  1758   
Albania         0     0     0     1  ...  1223   856   702   560   716   561   
Algeria        69    63    44    69  ...  3626  4807  3623  4005  5393  4752   

             2011  2012  2013   Total  
Country                                
Afghanistan  2203  2635  2004  117278  
Albania       539   620   603   31398  
Algeria      4325  3774  4331  138878  

[3 rows x 38 columns]

Let's view the number of immigrants from Japan (row 87) for the following scenarios: 1. The full row data (all columns) 2. For year 2013 3. For years 1980 to 1985

In [27]:
# 1. the full row data (all columns)
canadaimmigration.loc['Japan']

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [29]:
# alternate methods
canadaimmigration.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [30]:
canadaimmigration[canadaimmigration.index == 'Japan']

Continent        Region            DevName  1980  1981  1982  1983  \
Country                                                                      
Japan        Asia  Eastern Asia  Developed regions   701   756   598   309   

         1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  2011  \
Country                    ...                                             
Japan     246   198   248  ...  1067  1212  1250  1284  1194  1168  1265   

         2012  2013  Total  
Country                     
Japan    1214   982  55414  

[1 rows x 38 columns]

Check for zambia all the details in the immigration.

In [31]:
canadaimmigration[canadaimmigration.index=='Zambia']

Continent          Region             DevName  1980  1981  1982  1983  \
Country                                                                         
Zambia     Africa  Eastern Africa  Developing regions    11    17    11     7   

         1984  1985  1986  ...  2005  2006  2007  2008  2009  2010  2011  \
Country                    ...                                             
Zambia     16     9    15  ...    91    77    71    64    60   102    69   

         2012  2013  Total  
Country                     
Zambia     46    59   3354  

[1 rows x 38 columns]

In [32]:
# alternate method
# year 2013 is the last column, with a positional index of 36
canadaimmigration.iloc[87, 36]

982

In [35]:
# 3. for years 1980 to 1985
canadaimmigration.loc['Japan', ['1980', '1981', '1982', '1983', '1984', '1985']]

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object

Lets do Yemen Migration from the years 1980-1984.

In [36]:
canadaimmigration.loc['Yemen',['1980','1981','1982','1983','1984','1985']]

1980     1
1981     2
1982     1
1983     6
1984     0
1985    18
Name: Yemen, dtype: object

In [37]:
# Alternative Method
canadaimmigration.iloc[87, [3, 4, 5, 6, 7, 8]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Let me calculate for Algeria

In [39]:
canadaimmigration.iloc[2,[3,4,5,6,7,8]]

1980    80
1981    67
1982    71
1983    69
1984    63
1985    44
Name: Algeria, dtype: object

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.

In [44]:
canadaimmigration.columns = list(map(str, canadaimmigration.columns))
# [print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:

In [45]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

# <h3>Filtering based on a criteria.</h3>

To filter the dataframe based on a condition, we simply pass the condition as a boolean vector.

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).

In [46]:
# 1. create the condition boolean series
condition = canadaimmigration['Continent'] == 'Asia'
print(condition)

Country
Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [47]:
# 2. pass this condition into the dataFrame
canadaimmigration[condition]

Continent              Region  \
Country                                                                        
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                          

In [50]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

canadaimmigration[(canadaimmigration['Continent']=='Asia') & (canadaimmigration['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Continent         Region             DevName  1980  \
Country                                                                         
Afghanistan                     Asia  Southern Asia  Developing regions    16   
Bangladesh                      Asia  Southern Asia  Developing regions    83   
Bhutan                          Asia  Southern Asia  Developing regions     0   
India                           Asia  Southern Asia  Developing regions  8880   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions  1172   
Maldives                        Asia  Southern Asia  Developing regions     0   
Nepal                           Asia  Southern Asia  Developing regions     1   
Pakistan                        Asia  Southern Asia  Developing regions   978   
Sri Lanka                       Asia  Southern Asia  Developing regions   185   

                            1981  1982  1983  1984  1985  1986  ...   2005  \
Country                                                         ...          
Afghanistan                   39    39    47    71   340   496  ...   3436   
Bangladesh                    84    86    81    98    92   486  ...   4171   
Bhutan                         0     0     0     1     0     0  ...      5   
India                       8670  8147  7338  5704  4211  7150  ...  36210   
Iran (Islamic Republic of)  1429  1822  1592  1977  1648  1794  ...   5837   
Maldives                       0     0     1     0     0     0  ...      0   
Nepal                          1     6     1     2     4    13  ...    607   
Pakistan                     972  1201   900   668   514   691  ...  14314   
Sri Lanka                    371   290   197  1086   845  1838  ...   4930   

                             2006   2007   2008   2009   2010   2011   2012  \
Country                                                                       
Afghanistan                  3009   2652   2111   1746   1758   2203   2635   
Bangladesh                   4014   2897   2939   2104   4721   2694   2640   
Bhutan                         10      7     36    865   1464   1879   1075   
India                       33848  28742  28261  29456  34235  27509  30933   
Iran (Islamic Republic of)   7480   6974   6475   6580   7477   7479   7534   
Maldives                        0      2      1      7      4      3      1   
Nepal                         540    511    581    561   1392   1129   1185   
Pakistan                    13127  10124   8994   7217   6811   7468  11227   
Sri Lanka                    4714   4123   4756   4547   4422   3309   3338   

                             2013    Total  
Country                                     
Afghanistan                  2004   117278  
Bangladesh                   3789   131136  
Bhutan                        487    11752  
India                       33087  1383808  
Iran (Islamic Republic of)  11291   351846  
Maldives                        1       60  
Nepal                        1308    20444  
Pakistan                    12603   483200  
Sri Lanka                    2394   296716  

[9 rows x 38 columns]

In [51]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

canadaimmigration[(canadaimmigration['Continent']=='Africa') & (canadaimmigration['Region']=='Eastern Africa')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Continent          Region             DevName  \
Country                                                                     
Burundi                        Africa  Eastern Africa  Developing regions   
Comoros                        Africa  Eastern Africa  Developing regions   
Djibouti                       Africa  Eastern Africa  Developing regions   
Eritrea                        Africa  Eastern Africa  Developing regions   
Ethiopia                       Africa  Eastern Africa  Developing regions   
Kenya                          Africa  Eastern Africa  Developing regions   
Madagascar                     Africa  Eastern Africa  Developing regions   
Malawi                         Africa  Eastern Africa  Developing regions   
Mauritius                      Africa  Eastern Africa  Developing regions   
Mozambique                     Africa  Eastern Africa  Developing regions   
Rwanda                         Africa  Eastern Africa  Developing regions   
Seychelles                     Africa  Eastern Africa  Developing regions   
Somalia                        Africa  Eastern Africa  Developing regions   
Uganda                         Africa  Eastern Africa  Developing regions   
United Republic of Tanzania    Africa  Eastern Africa  Developing regions   
Zambia                         Africa  Eastern Africa  Developing regions   
Zimbabwe                       Africa  Eastern Africa  Developing regions   

                             1980  1981  1982  1983  1984  1985  1986  ...  \
Country                                                                ...   
Burundi                         0     0     0     0     1     2     1  ...   
Comoros                         0     2     2     0     0     2     1  ...   
Djibouti                        0     1     1     5     1     0     4  ...   
Eritrea                         0     0     0     0     0     0     0  ...   
Ethiopia                       98    59    54    98   154   157   302  ...   
Kenya                         291   296   221   201   218   245   244  ...   
Madagascar                     22    10    13    32    30    18    10  ...   
Malawi                          5     4     6     3     2     0     4  ...   
Mauritius                     306   259   353   170   207   164   325  ...   
Mozambique                      0     0     7     2     3     1     3  ...   
Rwanda                          6     6     8     4     3     0     9  ...   
Seychelles                      5    15     3     4     1     8    13  ...   
Somalia                         7    10     7    12    14    10    37  ...   
Uganda                         13    16    17    38    32    29    40  ...   
United Republic of Tanzania   635   832   621   474   473   460   403  ...   
Zambia                         11    17    11     7    16     9    15  ...   
Zimbabwe                       72   114   102    44    32    29    43  ...   

                             2005  2006  2007  2008  2009  2010  2011  2012  \
Country                                                                       
Burundi                       626   468   614   448   566   529   604   684   
Comoros                         5     7    11     3    19    22    14    18   
Djibouti                       93    68    50    47    70   123   127   164   
Eritrea                       528   700   699   799   904   931  1182  1335   
Ethiopia                     1506  1801  1512  1613  1289  1865  2163  1864   
Kenya                         693   622   449   453   489   507   585   725   
Madagascar                     88    96   115   118   169   178   148   165   
Malawi                         23    15    13     5    10    28    16     9   
Mauritius                     696   512   507   714   872  1455  1146   799   
Mozambique                      5     9     5     5    12     7     3     4   
Rwanda                        378   360   420   358   415   396   492   562   
Seychelles                     25     8     9    18   